lib/terminus.js

```diff
@@ -1,143 +1,171 @@
 'use strict'
 
 const stoppable = require('stoppable')
-const { promisify } = require('es6-promisify')
-
-const SUCCESS_RESPONSE = JSON.stringify({
-  status: 'ok'
-})
-
-const FAILURE_RESPONSE = JSON.stringify({
-  status: 'error'
-})
+const {
+  promisify
+} = require('es6-promisify')
+const { HealthCheckNotFoundError } = require('./terminus-error')
 
 function noopResolves () {
   return Promise.resolve()
 }
 
-async function sendSuccess (res, options) {
-  const { info } = options
+function noop () { }
+
+const DEFAULT_OPTIONS = {
+  signal: 'SIGTERM',
+  signals: [],
+  timeout: 1000,
+  healthChecks: {},
+  onShutdown: noopResolves,
+  beforeShutdown: noopResolves,
+  onSendFailureDuringShutdown: noopResolves,
+  logger: noop
+}
+
+/**
+ * A Terminus instance represents a handler
+ * for health checks and system signals
+ */
+class Terminus {
+  /**
+   * Initializes the Terminus instance by listening on
+   * system signals and decorating the given server with
+   * the given health checks.
+   * @param {any} server The HTTP server instance which should get decorated with the given healthchecks
+   * @param {object} options The options which configure the Termninus instance
+   */
+  constructor (server, options = {}) {
+    this._server = server
+    this._options = Object.assign(DEFAULT_OPTIONS, options)
+    this._state = {
+      isShuttingDown: false
+    }
+
+    this._onSignal = options.onSignal || options.onSigterm || noopResolves
 
-  res.statusCode = 200
-  res.setHeader('Content-Type', 'application/json')
-  if (info) {
-    return res.end(JSON.stringify({
-      status: 'ok',
-      info: info
-    }))
+    if (Object.keys(this._options.healthChecks).length > 0) {
+      this._decorateWithHealthCheck()
+    }
+
+    // push the signal into the array
+    // for backwards compatability
+    if (!this._options.signals.includes(this._options.signal)) {
+      this._options.signals.push(this._options.signal)
+    }
+    this._decorateWithSignalHandler()
   }
-  res.end(SUCCESS_RESPONSE)
-}
 
-async function sendFailure (res, options) {
-  const { error, onSendFailureDuringShutdown } = options
+  async _cleanup (signal) {
+    if (!this._state.isShuttingDown) {
+      this._state.isShuttingDown = true
+      try {
+        await this._options.beforeShutdown()
+        await this._asyncServerStop()
+        await this._onSignal()
+        await this._options.onShutdown()
+        this._options.signals.forEach(sig => process.removeListener(sig, () => this._cleanup()))
+        process.kill(process.pid, signal)
+      } catch (error) {
+        this._options.logger('error happened during shutdown', error)
+        process.exit(1)
+      }
+    }
+  }
 
-  if (onSendFailureDuringShutdown) {
-    await onSendFailureDuringShutdown()
+  _sendSuccess (info) {
```

Comment:
this name is now misleading, as this function won't send any success message, just generate one - can you pick a more descriptive name?